src/main/java/org/springframework/retry/annotation/JPARepositoryRetryBeanPostProcessor.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.retry.annotation;
+
+import org.springframework.aop.Advisor;
+import org.springframework.aop.framework.Advised;
+import org.springframework.aop.framework.ProxyFactory;
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.config.BeanPostProcessor;
+import org.springframework.core.Ordered;
+import org.springframework.data.jpa.repository.support.JpaRepositoryImplementation;
+import org.springframework.retry.interceptor.Retryable;
+
+/**
+ * Add a bean of this instance if you wish to annotate JPA Repository interfaces with
+ * {@code @Retryable}. It merges the repository proxy advisors with the retry advisor.
+ *
+ * @author Gary Russell
+ * @since 1.3.2
+ */
+public class JPARepositoryRetryBeanPostProcessor implements BeanPostProcessor, Ordered {
+
+	@Override
+	public int getOrder() {
+		return Integer.MAX_VALUE;
+	}
+
+	@Override
+	public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {
+		if (bean instanceof Retryable && bean instanceof Advised) {
```

Comment:
You could generally check if the bean implements `org.springframework.data.repository.Repository` (as marker for Spring Data repositories) to make this work for all Spring Data modules. Not sure why JpaRepositoryImplementation` needs to be excluded here.

Reply:
Thanks Mark; without removing that interface, I get

```
Caused by: java.lang.IllegalStateException: No MethodInvocation found: Check that an AOP invocation is in progress, and that the CrudMethodMetadataPopulatingMethodInterceptor is upfront in the interceptor chain.
	at org.springframework.data.jpa.repository.support.CrudMethodMetadataPostProcessor$CrudMethodMetadataPopulatingMethodInterceptor.currentInvocation(CrudMethodMetadataPostProcessor.java:129) ~[spring-data-jpa-2.5.1.jar:2.5.1]
	at org.springframework.data.jpa.repository.support.CrudMethodMetadataPostProcessor$ThreadBoundTargetSource.getTarget(CrudMethodMetadataPostProcessor.java:322) ~[spring-data-jpa-2.5.1.jar:2.5.1]
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:195) ~[spring-aop-5.3.7.jar:5.3.7]
```

I am going to close this PR for now, and open an issue against spring-aop, to see if we can get the root cause fixed.

Reply:
https://github.com/spring-projects/spring-framework/issues/27044

Let's see what the framework guys say before doing any more.